### PR TITLE
Enhance Literal class

### PR DESCRIPTION
**Pull request by kripken**
Work for #177.

First patch does Unary integer stuff. I'd like feedback on this before doing more work so I know I'm in the right direction.

### CODE REVIEW

**Comment by kripken** [comment]
cc @jfbastien

**Comment by jfbastien** [comment]
That looks like a great start!

I'd change most methods to be `const`, but otherwise lgtm.

**Comment by kripken** [comment]
Added constiness.

Also added a commit with float unaries. todo: binaries..

**Comment by kripken** [comment]
Refactoring complete.

**Comment by kripken** [comment]
Looks ok on travis. Should probably squash though since I have a fix commit at the end.

File: src/wasm-interpreter.h [hunk 1]
```diff
@@ -454,167 +444,104 @@ class ModuleInstance {
...
-            case Mul:      return Literal(l * r);
+            case Add:      return left.add(right);
+            case Sub:      return left.sub(right);
+            case Mul:      return left.mul(right);
             case DivS: {
-              if (r_signed == 0) trap("i32.div_s by 0");
-              if (l_signed == INT32_MIN && r_signed == -1) trap("i32.div_s overflow"); // signed division overflow
-              return Literal(l_signed / r_signed);
+              if (right.getInteger() == 0) trap("i32.div_s by 0");
+              if (left.getInteger() == INT32_MIN && right.getInteger() == -1) trap("i32.div_s overflow"); // signed division overflow
```
Comment (jfbastien) [inline src/wasm-interpreter.h]: Here and other places, could you use `numeric_limits` instead of the macros?

File: src/wasm.h [hunk 2]
```diff
@@ -273,6 +286,382 @@ class Literal {
...
+    assert(type == WasmType::f32);
+    return Literal(double(getf32()));
+  }
+  Literal truncateToI32() const {
+    assert(type == WasmType::i64);
+    return Literal((int32_t)i64);
+  }
+  Literal truncateToF32() const {
+    assert(type == WasmType::f64);
+    return Literal(float(getf64()));
```
Comment (jfbastien) [inline src/wasm.h]: What do both truncates to when out of range? Are there tests for them?

File: src/wasm.h [hunk 3]
```diff
@@ -273,6 +286,382 @@ class Literal {
...
+    if (type == WasmType::i32) return Literal(double(uint32_t(i32)));
+    if (type == WasmType::i64) return Literal(double(uint64_t(i64)));
+    WASM_UNREACHABLE();
+  }
+
+  Literal neg() const {
+    switch (type) {
+      case WasmType::i32: return Literal(i32 ^ 0x80000000);
+      case WasmType::i64: return Literal(int64_t(i64 ^ 0x8000000000000000ULL));
+      case WasmType::f32: return Literal(i32 ^ 0x80000000).castToF32();
```
Comment (jfbastien) [inline src/wasm.h]: The active member is `f32` here, no? That would need to be `bit_cast` to `i32` first.

File: src/wasm.h [hunk 4]
```diff
@@ -273,6 +286,382 @@ class Literal {
...
+      case WasmType::f64: return Literal(int64_t(i64 ^ 0x8000000000000000ULL)).castToF64();
```
Comment (jfbastien) [inline src/wasm.h]: Same.

File: src/wasm.h [hunk 5]
```diff
@@ -273,6 +286,382 @@ class Literal {
...
+    switch (type) {
+      case WasmType::f32: return Literal(std::nearbyint(getf32()));
+      case WasmType::f64: return Literal(std::nearbyint(getf64()));
+      default: WASM_UNREACHABLE();
+    }
+  }
+  Literal sqrt() const {
+    switch (type) {
+      case WasmType::f32: return Literal(std::sqrt(getf32()));
+      case WasmType::f64: return Literal(std::sqrt(getf64()));
```
Comment (jfbastien) [inline src/wasm.h]: Here too, what happens on invalid inputs and is it tested?

File: src/wasm.h [hunk 6]
```diff
@@ -273,6 +286,382 @@ class Literal {
...
+      case WasmType::f64: return Literal(getf64() >= other.getf64());
+      default: WASM_UNREACHABLE();
+    }
+  }
+
+  Literal min(const Literal& other) const {
+    switch (type) {
+      case WasmType::f32: {
+        auto l = getf32(), r = other.getf32();
+        if (l == r && l == 0) return Literal(1/l < 0 ? l : r);
```
Comment (jfbastien) [inline src/wasm.h]: `std::signbit`

File: src/wasm.h (hunk 2 above)
Comment (kripken) [inline src/wasm.h]: There are spec tests for this, yes. But I don't know offhand what they do, just that this passes them ;)

File: src/wasm.h (hunk 3 above)
Comment (kripken) [inline src/wasm.h]: No, it's already stored as just bits in the i32. We never store float values directly.

File: src/wasm.h (hunk 5 above)
Comment (kripken) [inline src/wasm.h]: Spec test suite has a lot of tests for this. Hopefully it covers all invalid inputs.

**Comment by kripken** [comment]
Addressed the review comments.

File: src/wasm.h (hunk 5 above)
Comment (jfbastien) [inline src/wasm.h]: I'm not super comfortable mapping the C++ STL behavior to wasm's, but I guess it'll do for now.

**Comment by jfbastien** [comment]
lgtm for now, though the corner cases worry me this is a clear improvement.

**Comment by kripken** [comment]
Cool, squashed and pushed.

I agree more testcases for corner cases would be good. The spec repo already has a lot, but honestly I didn't look carefully to see if they are complete or not. @sunfishcode has been interested in that though I believe.

File: src/wasm.h (hunk 5 above)
Comment (sunfishcode) [inline src/wasm.h]: Standard libary `sqrt` functions are pretty reliable in practice.

**Pull request closed without merging**